lib/modules/versioning/deb/index.ts

```diff
@@ -0,0 +1,145 @@
+import { GenericVersion, GenericVersioningApi } from '../generic';
+import type { VersioningApi } from '../types';
+
+export const id = 'deb';
+export const displayName = 'Deb version';
+export const urls = [
+  'https://www.debian.org/doc/debian-policy/ch-controlfields.html#version',
+  'https://manpages.debian.org/unstable/dpkg-dev/deb-version.7.en.html',
+];
+export const supportsRanges = false;
+
+const epochPattern = regEx(/^\d+$/);
+const upstreamVersionPattern = regEx(/^[-+.:~A-Za-z\d]+$/);
+const debianRevisionPattern = regEx(/^[+.~A-Za-z\d]*$/);
+const numericPattern = regEx(/\d+/g);
+const characterOrder =
+  '~ ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz+-.:';
+const numericChars = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
+
+export interface DebVersion extends GenericVersion {
+  /**
+   * epoch, defaults to 0 if not present, are used to leave version mistakes and previous
+   * versioning schemes behind.
+   */
+  epoch: number;
+  /**
+   * upstreamVersion is the main version part: it defines the version of origin software
+   * that was packaged.
+   */
+  upstreamVersion: string;
+  /**
+   * debianRevision is used to distinguish between different versions of packaging for the
+   * same upstream version.
+   */
+  debianRevision: string;
+}
+
+class DebVersioningApi extends GenericVersioningApi {
+  protected _parse(version: string): DebVersion | null {
+    /* Splitting version into "[epoch:]upstream-version[-debian-revision]"
+       All found numbers are exported as release info */
+
+    // split of first element by `:` as epoch:
+    const epochSplit = version.split(':');
+    const epochStr = epochSplit.length > 1 ? epochSplit.shift()! : '0';
+    const remainingVersion = epochSplit.join(':');
+
+    // split of last element by `-`
+    if (remainingVersion.endsWith('-')) {
+      // Forbid debian revision (it would result in `2.4.0-` == `2.4.0`)
+      return null;
+    }
+    const debianSplit = remainingVersion.split('-');
+    const debianRevision = debianSplit.length > 1 ? debianSplit.pop()! : '';
+    const upstreamVersion = debianSplit.join('-');
+
+    if (
+      !epochPattern.test(epochStr) ||
+      !upstreamVersionPattern.test(upstreamVersion) ||
+      !debianRevisionPattern.test(debianRevision)
+    ) {
+      return null;
+    }
+    const release = [...remainingVersion.matchAll(numericPattern)].map((m) =>
+      parseInt(m[0], 10)
+    );
+    return {
+      epoch: parseInt(epochStr, 10),
+      upstreamVersion,
+      debianRevision,
+      release,
+      suffix: debianRevision,
+    };
+  }
+
+  protected _compare_string(a: string, b: string): number {
+    /* The string is compared by continuous blocks of a) non-digit and b) digit characters.
+       Non-digit blocks are compared lexicographically with a custom character order.
+       Digit blocks are compared numerically.
+    */
+    let charPos = 0;
+    while (charPos < a.length || charPos < b.length) {
```

Comment:
this loop looks pretty complicated. can't we just do a numerical locale compare on the string?

that should be smart enough to detect `a10b` as greater than `a9b`. 🤔

@zharinov WDYT?

Reply:
A numerical locale compare is almost correct. However, the [spec](https://manpages.debian.org/unstable/dpkg-dev/deb-version.7.en.html#Sorting_algorithm) requires smaller sorting priority for tilde and empty space: 

> The lexical comparison is a comparison of ASCII values modified so that all the letters sort earlier than all the non-letters and so that a tilde sorts before anything, even the end of a part. For example, the following parts are in sorted order: **‘~~’, ‘~~a’, ‘~’, the empty part, ‘a’**.

The function implementation follows the documented behavior (alternating between sorting non-digit and digit blocks within the strings). However, it should be possible to simplify the algorithm to just look at the first non-matching character (and decided based on surrounding characters and the character preferences what to do).

Reply:
Can you add a comment with the link where the algorithm is documented.

Reply:
Done. FYI the simplified implementation (maybe a bit shorter, maybe easier to understand?):

```ts
  protected _compare_string(a: string, b: string): number {
    /* Special string sorting based on official specification:
     * https://manpages.debian.org/unstable/dpkg-dev/deb-version.7.en.html#Sorting_algorithm
     * The string is compared by continuous blocks of a) non-digit and b) digit characters.
     * Non-digit blocks are compared lexicographically with a custom character order.
     * Digit blocks are compared numerically.
     * This algorithm is simplified by searching the first non-matching character and using the
     * surrounding characters to decided whether to compare numerically (digit characters)
     * or lexicographically (non-digits)
     */
    let firstMismatch = 0;
    while (
      a.charAt(firstMismatch) === b.charAt(firstMismatch) &&
      firstMismatch < a.length
    ) {
      firstMismatch += 1;
    }
    const aChar = a.charAt(firstMismatch);
    const bChar = b.charAt(firstMismatch);

    const previouslyNumbers = numericChars.includes(
      a.charAt(firstMismatch - 1)
    );
    const aIsNumber = numericChars.includes(aChar);
    const bIsNumber = numericChars.includes(bChar);
    if (
      (previouslyNumbers && (aIsNumber || bIsNumber)) ||
      (aIsNumber && bIsNumber)
    ) {
      // Compare numerically (we still are in a digit block or a new starts here)
      return a.localeCompare(b, undefined, { numeric: true });
    } else {
      // Compare lexicographically
      // Numbers are treated like end of string ('', they would be part of the next digit block)
      // Both represented as blank character during ordering
      const aPriority = characterOrder.indexOf(
        aIsNumber || aChar === '' ? ' ' : aChar
      );
      const bPriority = characterOrder.indexOf(
        bIsNumber || bChar === '' ? ' ' : bChar
      );
      return Math.sign(aPriority - bPriority);
    }
  }
```

Reply:
this looks much better